### PR TITLE
Auto Options Per Route

### PR DESCRIPTION
**Pull request by webRat**
This will take every incoming HandlerDefs and uniquely run through all of them taking note of all the Path/Methods and then create OPTIONS method for each unique Path.

So, if /organizations has a GET and a POST method, then hitting /organizations with OPTIONS should automatically return Allowed: "GET,POST". Same for /organizations/:orgID, etc.

*WARNING*: This also includes code from previous pull request.

### CODE REVIEW

File: webserver/webserver.go [hunk 1]
```diff
@@ -299,3 +318,15 @@ func (s *Server) onMissingHandler(w http.ResponseWriter, req *http.Request) {
...
 	}
 }
+
+func createOption(path string, methods []string) HandlerDef {
+	return HandlerDef{
+		Alias:  path + "OptionsCheck",
+		Method: "OPTIONS",
+		Path:   path,
+		Handler: func(c *context.Context) {
+			c.Output.Header("Allowed", strings.Join(methods, ","))
```
Comment (aarongreenlee) [inline webserver/webserver.go]: @webRat, I think the HandlerDef needs to be modified to allow the handler to specify incoming headers as well. Currently only outgoing headers are listed. Any incoming headers would be transmitted in the OPTIONS response with the header `access-control-allow-headers`. More reading: https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Access-Control-Allow-Headers

**Comment by webRat** [comment]
Working on a fix.

File: webserver/webserver.go (hunk 1 above)
Comment (webRat) [inline webserver/webserver.go]: Working on a fix.

File: webserver/webserver.go [hunk 2]
```diff
@@ -158,6 +158,25 @@ func New(
 	return s
 }
 
+// RegisterHandlerDefsWithOptions accepts a slice of HandlerDefs and registers
+// each unique route and then after all the routes have been determined, creates
+// a new handler def with the OPTIONS method.
+func (s *Server) RegisterHandlerDefsWithOptions(h []HandlerDef) {
```
Comment (aarongreenlee) [inline webserver/webserver.go]: @webRat, thinking about the desired behavior, I wonder if `RegisterHandlerDefsAndOptions` is a better description of what we expect to accomplish with this method. What do you think?

Comment (webRat) [inline webserver/webserver.go]: This isn't the most recent commit. It's on my machine and I wanted to review with you before I commited.

**Comment by webRat** [comment]
Updated based on feedback.

**Comment by aarongreenlee** [comment]
Thanks, @webRat!

**Pull request merged**